backend/job_tracker.py

```diff
@@ -0,0 +1,332 @@
+# Copyright OpenSearch Contributors
+# SPDX-License-Identifier: Apache-2.0
+
+import json
+import os
+import threading
+import uuid
+from datetime import datetime
+from typing import Dict, Any, Optional
+from enum import Enum
+import logging
+from constants import ErrorMessages
+
+logger = logging.getLogger(__name__)
+
+# Job status constants
+class JobStatus(Enum):
+    QUEUED = "queued"
+    RUNNING = "running"
+    COMPLETED = "completed"
+    FAILED = "failed"
+    CANCELLED = "cancelled"
+
+# Task status constants
+class TaskStatus(Enum):
+    PENDING = "pending"
+    RUNNING = "running"
+    COMPLETED = "completed"
+    FAILED = "failed"
+    SKIPPED = "skipped"
+
+# Job tracker class
+class JobTracker:
+    """Simple job tracking system with file-based persistence."""
+    
+    def __init__(self, jobs_dir: str = "job_data"):
+        # Convert to absolute path immediately to be immune to working directory changes
+        self.jobs_dir = os.path.abspath(jobs_dir)
+        self.lock = threading.Lock()
+        
+        # Create jobs directory if it doesn't exist
+        os.makedirs(self.jobs_dir, exist_ok=True)
+        
+        # Clean up any jobs that were running when server was restarted
+        self.cleanup_interrupted_jobs()
+        
+        logger.info(f"JobTracker initialized with absolute jobs directory: {self.jobs_dir}")
+    
+    # cleanup the interrupted jobs
+    def cleanup_interrupted_jobs(self) -> None:
+        """Clean up jobs that were running when the server was restarted."""
+        try:
+            cleaned_jobs = 0
+            for filename in os.listdir(self.jobs_dir):
+                if filename.endswith('.json'):
+                    job_id = filename[:-5]
+                    job_data = self.load_job(job_id)
+                    if job_data and self.is_job_interrupted(job_data):
+                        self.mark_job_as_interrupted(job_id, job_data)
+                        cleaned_jobs += 1
+            
+            if cleaned_jobs > 0:
+                logger.info(f"Cleaned up {cleaned_jobs} interrupted jobs from previous server session")
+        except Exception as e:
+            logger.error(f"Failed to cleanup interrupted jobs: {e}")
+
+    # check if a job was interrupted
+    def is_job_interrupted(self, job_data: Dict[str, Any]) -> bool:
+        """Check if a job was interrupted (running/queued when server stopped)."""
+        job_status = job_data.get("status")
+        
+        if job_status in ["running", "queued"]:
```

Comment:
This logic assumes any "running" or "queued" job was interrupted, but what if the server restarts while jobs are legitimately running? Consider adding a timestamp-based check or process ID validation. You could later enhance it with add TODO here.